### PR TITLE
Add hooking support for Unity apps

### PR DESCRIPTION
**Pull request by rjoudrey**

### CODE REVIEW

**Comment by LibreVR** [comment]
Interesting, do we have any idea why hooking LoadLibrary fails, but hooking GetProcAddress does work? It doesn't really make sense.

Also, instead of using a map, wouldn't it be easier to just call TrueGetProcAdress with our module handle?

File: main.cpp [hunk 1]
```diff
@@ -47,20 +125,28 @@ HANDLE WINAPI HookOpenEvent(DWORD dwDesiredAccess, BOOL bInheritHandle, LPCWSTR
 BOOL APIENTRY DllMain(HANDLE hModule, DWORD ul_reason_for_call, LPVOID lpReserved)
 {
 	ReviveModule = hModule;
```
Comment (LibreVR) [inline main.cpp]: Perhaps if we move this line to `DLL_PROCESS_ATTACH` then maybe it won't be overwritten by an invalid handle.

File: main.cpp [hunk 2]
```diff
@@ -13,27 +19,99 @@
...
 
+bool isUnityGame = false;
+HMODULE WINAPI HookLoadLibrary(LPCWSTR lpFileName)
+{
 	LPCWSTR fileName = PathFindFileNameW(lpFileName);
-	WCHAR libraryName[MAX_PATH];
-	swprintf(libraryName, MAX_PATH, L"LibOVRRT%hs_%d.dll", pBitDepth, OVR_MAJOR_VERSION);
-	if (wcscmp(fileName, libraryName) == 0)
+	if (!isUnityGame && wcscmp(fileName, libraryName) == 0)
 		return (HMODULE)ReviveModule;
```
Comment (LibreVR) [inline main.cpp]: Or we could also call `GetModuleHandle()` with our own DLL name here, maybe we'll return a valid handle then.

**Comment by rjoudrey** [comment]
The reason why hooking LoadLibrary fails in Unity is because there are security checks being done on the module returned from LoadLibrary.

**Comment by LibreVR** [comment]
> The reason why hooking LoadLibrary fails in Unity is because there are security checks being done on the module returned from LoadLibrary.

But we're only replacing the module when OVRPlugin.dll loads it, we don't touch any module loaded by Unity.

File: main.cpp [hunk 3]
```diff
@@ -13,27 +19,99 @@
...
-#if defined(_WIN64)
-	const char* pBitDepth = "64";
-#else
-	const char* pBitDepth = "32";
-#endif
+	WCHAR modulePath[MAX_PATH];
+	GetModuleFileName(hModule, modulePath, sizeof(modulePath));
+	LPCWSTR moduleName = PathFindFileNameW(modulePath);
+	if (wcscmp(moduleName, libraryName) == 0) {
+		auto iter = lookup.find(lpProcName);
```
Comment (LibreVR) [inline main.cpp]: Instead of looking through a map, we could just call `TrueGetProcAddress(ReviveModule, lpProcName)` here.

**Comment by rjoudrey** [comment]
OVRPlugin's code for loading LibOVR basically looks like this:
lib = LoadLibrary(LibOVR)
doSecurityChecks(lib)
if (securityChecksFailed()) {

}

**Comment by LibreVR** [comment]
That is indeed more secure, but then my patched `OVRPlugin.dll` should also fail.

**Comment by rjoudrey** [comment]
The check's aren't being done on OVRPlugin.dll

**Comment by LibreVR** [comment]
@rjoudrey I mean when I patch `OVRPlugin.dll` I only force it to load LibRevive, I still rely on HookLoadLibrary to fool `OVRPlugin.dll` into using the Revive library.

**Comment by rjoudrey** [comment]
I understand, and I believe that this is because GetProcAddress may actually try to read from the current module's IAT if possible.

**Comment by LibreVR** [comment]
@rjoudrey `OVRPlugin.dll` load the library at runtime, LibOVR is not part of its IAT. I really think we're returning an invalid module handle, could you try some of the suggestions.

If neither of my suggestions work, then I think we should just get rid of HookLibrary altogether and always use HookGetProcAddress.

**Comment by rjoudrey** [comment]
GetProcAddress doesn't get called for Unreal Engine games, since those games have LibOVR in their IAT.

**Comment by LibreVR** [comment]
@rjoudrey But if you have LibOVR in your IAT, then Windows will load the DLL right? Then you would never call `LoadLibrary`.

**Comment by rjoudrey** [comment]
I don't see LibOVR in Dreamdeck's imports. So I think you may be correct about getting rid of LoadLibrary altogether.

**Comment by rjoudrey** [comment]
I've removed the map and LoadLibrary hook, and it seems to work in Lucky's Tale and Dreamdeck.

**Comment by LibreVR** [comment]
Interesting, so the Revive module handle is actually still valid. I still want to do a few more tests with `HookLoadLibrary` out of curiosity. But it seems likely that hooking `GetProcAddress` is more reliable.

File: main.cpp [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #include <stdio.h>
 #include "MinHook.h"
 #include "Shlwapi.h"
+#include <map>
```
Comment (LibreVR) [inline main.cpp]: You don't need to include the map header anymore.

File: main.cpp [hunk 5]
```diff
@@ -47,20 +43,25 @@ HANDLE WINAPI HookOpenEvent(DWORD dwDesiredAccess, BOOL bInheritHandle, LPCWSTR
 BOOL APIENTRY DllMain(HANDLE hModule, DWORD ul_reason_for_call, LPVOID lpReserved)
 {
 	ReviveModule = hModule;
-
+#if defined(_WIN64)
+	const char* pBitDepth = "64";
+#else
+	const char* pBitDepth = "32";
+#endif
+	swprintf(libraryName, MAX_PATH, L"LibOVRRT%hs_%d.dll", pBitDepth, OVR_MAJOR_VERSION);
```
Comment (LibreVR) [inline main.cpp]: Indenting is misaligned here.

File: main.cpp [hunk 6]
```diff
@@ -47,20 +43,25 @@ HANDLE WINAPI HookOpenEvent(DWORD dwDesiredAccess, BOOL bInheritHandle, LPCWSTR
...
     switch(ul_reason_for_call)
     {
         case DLL_PROCESS_ATTACH:
 			MH_Initialize();
-			MH_CreateHook(LoadLibraryW, HookLoadLibrary, (PVOID*)&TrueLoadLibrary);
 			MH_CreateHook(OpenEventW, HookOpenEvent, (PVOID*)&TrueOpenEvent);
-			MH_EnableHook(LoadLibraryW);
+			MH_CreateHook(GetProcAddress, HookGetProcAddress, (PVOID*)&TrueGetProcAddress);
 			MH_EnableHook(OpenEventW);
+			MH_EnableHook(GetProcAddress);
```
Comment (LibreVR) [inline main.cpp]: Don't forget to also update the MH_EnableHook/MH_DisableHook in REV_CAPI.cpp.

**Comment by rjoudrey** [comment]
The behavior I've observed with Unity is that LibRevive is detached from the process very soon after returning the handle to it in HookLoadLibrary (due to security check failures I believe).

**Comment by LibreVR** [comment]
@rjoudrey Interesting, have you tried hooking `FreeLibrary` and ignore any instance where Unity tries to call it with our module handle?

**Comment by LibreVR** [comment]
The reason `HookLoadLibrary` fails is because we don't keep a reference count on our own module handle. We should either increment our reference count by calling `LoadLibrary` or implement our own reference count and hook `FreeLibrary`.

I think that's a cleaner solution than hooking `GetProcAddress`.

**Comment by rjoudrey** [comment]
Have you tested that?

**Comment by LibreVR** [comment]
@rjoudrey No, but it is consistent with all our findings, and the fact that `LoadLibrary` is expected to do reference counting is well-documented.

**Comment by LibreVR** [comment]
We should probably just not do reference counting and just ignore `FreeLibrary` altogether. Since there's no way our library will get loaded again once the reference count reaches zero.

**Comment by rjoudrey** [comment]
Looking over the disassembly again, it I was definitely wrong. The certificate checks are being done on the LibOVR file handle, before LoadLibrary gets called. CloseHandle is being called on the LibOVR file handle after LoadLibrary finishes. Have a look at text:000000018003B750 in OVRPlugin.dll to see what I mean.

**Comment by LibreVR** [comment]
@rjoudrey Yeah, so what's happening is that Unity and OVRPlugin both call LoadLibrary for LibOVR, then Unity frees that library because it doesn't need it anymore and OVRPlugin is stuck with invalid function pointers.

**Comment by LibreVR** [comment]
I will close this PR, please submit one that hooks `FreeLibrary`, if that doesn't work we can reopen this one.

**Pull request closed without merging**